# Hands-on Lab: Monitoring and Optimizing Your Databases in PostgreSQL
https://www.coursera.org/learn/relational-database-administration/ungradedLti/aGpos/hands-on-lab-monitoring-and-optimizing-your-databases-in-postgresql

in this lab, you’ll learn how to monitor and optimize your database in PostgreSQL with both the command line interface (CLI) and database administration tool, pgAdmin.

## Objectives
After completing this lab, you will be able to:

* Monitor the performance of your database with the command line interface and pgAdmin.
* Identify optimal data types for your database.
* Optimize your database via the command line with best practices.


## Database Used in this Lab
In this lab, you will use a database from https://postgrespro.com/education/demodb distributed under the PostgreSQL licence. It stores a month of data about airline flights in Russia and is organized according to the following schema:

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/DB_schema.png" alt="DB Schema" width="800" height="500">


## Exercise 1: Create Your Database

In [10]:
#!curl -O https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/example-guided-project/flights_RUSSIA_small.sql

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 99.0M  100 99.0M    0     0   490k      0  0:03:26  0:03:26 --:--:--  125k5  0:00:32  0:07:23  411k2:17  0:02:05  0:00:12  100k  0:02:44  0:02:34  0:00:10 97424:56  0:00:06  111k 0:03:04  0:00:04  173k


Terminal CLI

psql --username=postgres --host=localhost 

\i flights_RUSSIA_small.sql

\dt


|  Schema  |      Name       |  Type |  Owner   |
|----------|-----------------|-------|----------|
| bookings | aircrafts_data  | table | postgres |
| bookings | airports_data   | table | postgres |
| bookings | boarding_passes | table | postgres |
| bookings | bookings        | table | postgres |
| bookings | flights         | table | postgres |
| bookings | seats           | table | postgres |
| bookings | ticket_flights  | table | postgres |
| bookings | tickets         | table | postgres |


In [14]:
import os
from dotenv import load_dotenv

# Cargar variables de entorno desde el archivo .env
load_dotenv()

# Obtener la contraseña de la variable de entorno
password = os.getenv("DB_PASSWORD")

In [15]:
# Load SQL Magic extension
%load_ext sql

In [17]:
# Especificar los detalles de la conexión
user = 'postgres'
host = 'localhost'
port = '5432'  # Puerto predeterminado de PostgreSQL
dbname = 'demo'  # Nombre de la base de datos

# Crear la URL de conexión
connection_string = f'postgresql://{user}:{password}@{host}:{port}/{dbname}'

# Conectar a la base de datos
%sql $connection_string

## Exercise 2: Monitor Your Database

Database monitoring refers to reviewing the operational status of your database and maintaining its health and performance. With proper and proactive monitoring, databases will be able to maintain a consistent performance. Any problems that emerge, such as sudden outages, can be identified and resolved in a timely manner.

### Task A: Monitor Current Activity
To start, let’s take a look at how you can monitor current server and database activity in PostgreSQL.

### Server Activity
You can take a look at the server activity by running the following query:

In [18]:
%sql SELECT pid, usename, datname, state, state_change FROM pg_stat_activity;

 * postgresql://postgres:***@localhost:5432/demo
9 rows affected.


[(690, None, None, None, None),
 (691, 'postgres', None, None, None),
 (58524, 'postgres', 'postgres', 'idle', datetime.datetime(2024, 4, 15, 16, 12, 54, 96580, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None))),
 (50024, 'postgres', 'demo', 'idle', datetime.datetime(2024, 4, 15, 17, 3, 17, 488671, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None))),
 (58571, 'postgres', 'demo', 'idle', datetime.datetime(2024, 4, 15, 16, 17, 28, 616135, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None))),
 (91825, 'postgres', 'demo', 'active', datetime.datetime(2024, 4, 15, 17, 37, 51, 757215, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None))),
 (682, None, None, None, None),
 (681, None, None, None, None),
 (689, None, None, None, None)]

This query will retrieve the following:

| Column      | Description                                                   |
|-------------|---------------------------------------------------------------|
| pid         | Process ID                                                    |
| usename     | Name of user logged in                                        |
| datname     | Name of database                                              |
| state       | Current state, with two common values being: active (executing a query) and idle (waiting for new command) |
| state_change| Time when the state was last changed                          |


|  pid  | usename  | datname  | state  |         state_change          |
|-------|----------|----------|--------|-------------------------------|
|   690 |   -      |    -     |   -    |           -                   |
|   691 | postgres |    -     |   -    |           -                   |
| 58524 | postgres | postgres | idle   | 2024-04-15 16:12:54.09658-04  |
| 50024 | postgres | demo     | active | 2024-04-15 16:19:49.058763-04 |
| 58571 | postgres | demo     | idle   | 2024-04-15 16:17:28.616135-04 |
|   682 |   -      |    -     |   -    |           -                   |
|   681 |   -      |    -     |   -    |           -                   |
|   689 |   -      |    -     |   -    |           -                   |


Let’s say you wanted to see all the aforementioned columns, in addition to the actual text of the query that was last executed. Which column should you add to review that?

In [19]:
%sql SELECT pid, usename, datname, state, state_change, query FROM pg_stat_activity;

 * postgresql://postgres:***@localhost:5432/demo
9 rows affected.


[(690, None, None, None, None, ''),
 (691, 'postgres', None, None, None, ''),
 (58524, 'postgres', 'postgres', 'idle', datetime.datetime(2024, 4, 15, 16, 12, 54, 96580, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None)), "SELECT\n    pg_catalog.split_part(rolconfig, '=', 1) AS name, pg_catalog.replace(rolconfig, pg_catalog.split_part(rolconfig, '=', 1) || '=', '') AS v ... (572 characters truncated) ...             dr.setrole=25018::OID AND dr.setdatabase!=0\n        ) c\n        LEFT JOIN pg_catalog.pg_database d ON (d.oid = c.setdatabase)\n    ) a;"),
 (50024, 'postgres', 'demo', 'idle', datetime.datetime(2024, 4, 15, 17, 3, 17, 488671, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None)), 'SELECT datname, tup_inserted, tup_updated, tup_deleted FROM pg_stat_database;'),
 (58571, 'postgres', 'demo', 'idle', datetime.datetime(2024, 4, 15, 16, 17, 28, 616135, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None)), 'SELECT\n    x.oid, pg_catalog.pg_get_userbyid(extowner) AS owner,\n    x.extname AS name, n.nspname AS schema,\n    x.extrelocatable AS relocatable,  ... (114 characters truncated) ... e n ON x.extnamespace=n.oid\n    JOIN pg_catalog.pg_available_extensions() e(name, default_version, comment) ON x.extname=e.name ORDER BY x.extname\n'),
 (91825, 'postgres', 'demo', 'active', datetime.datetime(2024, 4, 15, 17, 38, 18, 469442, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-240, name=None)), 'SELECT pid, usename, datname, state, state_change, query FROM pg_stat_activity;'),
 (682, None, None, None, None, ''),
 (681, None, None, None, None, ''),
 (689, None, None, None, None, '')]

This column returns the most recent query. If state is active, it’ll show the currently executed query. If not, it’ll show the last query that was executed.

Your result should look similar to the following:

View pg_stat_activity with pid, usename, datname, state, state_change, and query

Notice how for the demo database, with a status of active, the current query you are executing is the one listed in the query column.

|  pid  | usename  | datname  | state  |         state_change          |                                                                                   query                                                                                   |
|-------|----------|----------|--------|-------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|   690 |    -     |    -     |   -    |               -               |                                                                                         -                                                                                   |
|   691 | postgres |    -     |   -    |               -               |                                                                                         -                                                                                   |
| 58524 | postgres | postgres | idle   | 2024-04-15 16:12:54.09658-04  | SELECT                                                                                                                                                                   +                                                                                                                |
|   -   |    -     |    -     |   -    |               -               |     pg_catalog.split_part(rolconfig, '=', 1) AS name, pg_catalog.replace(rolconfig, pg_catalog.split_part(rolconfig, '=', 1) || '=', '') AS value, NULL::text AS database+                                                                                                                |
|   -   |    -     |    -     |   -    |               -               | FROM                                                                                                                                                                     +                                                                                                                |
|   -   |    -     |    -     |   -    |               -               |     (SELECT                                                                                                                                                              +                                                                                                                |
|   -   |    -     |    -     |   -    |               -               |             pg_catalog.unnest(rolconfig) AS rolconfig, rolcanlogin, rolname                                                                                                |


Please note, if your table looks strange or squished, you can resize the terminal window by dragging it out.

If your result shows the text (END), then type in q to exit that view. Whenever you encounter this view, you can use q to return to your original view.

### Database Activity
When looking at database activity, you can use the following query:

In [20]:
%sql SELECT datname, tup_inserted, tup_updated, tup_deleted FROM pg_stat_database;

 * postgresql://postgres:***@localhost:5432/demo
7 rows affected.


[(None, 38, 5, 33),
 ('postgres', 0, 0, 0),
 ('coffee', 0, 0, 0),
 ('template1', 0, 0, 0),
 ('template0', 0, 0, 0),
 ('restored_demo', 0, 0, 0),
 ('demo', 2290150, 22, 0)]

This query will retrieve the following:

|   Column    |                           Description                            |
|-------------|------------------------------------------------------------------|
|  datname    |                      Name of database                            |
| tup_inserted|  Number of rows inserted by queries in this database            |
| tup_updated |  Number of rows updated by queries in this database             |
| tup_deleted |  Number of rows deleted by queries in this database             |


|    datname    | tup_inserted | tup_updated | tup_deleted |
|---------------|--------------|-------------|-------------|
|       -       |           38 |           5 |          33 |
| postgres      |            0 |           0 |           0 |
| coffee        |            0 |           0 |           0 |
| template1     |            0 |           0 |           0 |
| template0     |            0 |           0 |           0 |
| restored_demo |            0 |           0 |           0 |
| demo          |      2290150 |          22 |           0 |


As you can see, the two databases that are returned are the postgres and demo. These are databases that you are familiar with.

The other two, template1 and template0 are default templates for databases, and can be overlooked in this analysis.

Based on this output, you now know that demo had about 2,290,162 rows inserted and 22 rows updated.

To see what other columns are available for viewing, you can read through the pg_stat_database documentation.
https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-DATABASE-VIEW

Let’s say you wanted to see the number or rows fetched and returned by this database.

Note: The number of rows fetched is the number of rows that were returned. The number of rows returned is the number of rows that were read and scanned by the query.

What query should you use to do that?

In [21]:
%sql SELECT datname, tup_fetched, tup_returned FROM pg_stat_database;

 * postgresql://postgres:***@localhost:5432/demo
7 rows affected.


[(None, 70674, 199293),
 ('postgres', 72352, 3549216),
 ('coffee', 63126, 3654506),
 ('template1', 5093, 99417),
 ('template0', 0, 0),
 ('restored_demo', 64883, 3782293),
 ('demo', 355812, 7999070)]

Notice how the rows returned tend to be greater than the rows fetched. If you consider how tables are read, this makes sense because not all the rows scanned may be the ones that are returned.

With queries, you can apply filtering. What if you only wanted to see the database details (rows inserted, updated, deleted, returned and fetched) for demo?

In [22]:
%sql SELECT datname, tup_inserted, tup_updated, tup_deleted, tup_fetched, tup_returned FROM pg_stat_database WHERE datname = 'demo';

 * postgresql://postgres:***@localhost:5432/demo
1 rows affected.


[('demo', 2290150, 22, 0, 355840, 8000958)]

### Task B: Monitor Performance Over Time

Extensions, which can enhance your PostgreSQL experience, can be helpful in monitoring your database. One such extension is pg_stat_statements, which gives you an aggregated view of query statistics.

To enable the extension, enter the following command:

In [23]:
%sql CREATE EXTENSION pg_stat_statements;

 * postgresql://postgres:***@localhost:5432/demo
Done.


[]

This will enable the pg_stat_statements extension, which will start to track the statistics for your database.

Now, let’s edit the PostgreSQL configuration file to include the extension you just added:

In [30]:
%sql ALTER SYSTEM SET shared_preload_libraries = 'pg_stat_statements';

 * postgresql://postgres:***@localhost:5432/demo
(psycopg2.errors.ActiveSqlTransaction) ALTER SYSTEM cannot run inside a transaction block

[SQL: ALTER SYSTEM SET shared_preload_libraries = 'pg_stat_statements' ;]
(Background on this error at: https://sqlalche.me/e/20/2j85)


Execute in CLI :

#ALTER SYSTEM SET shared_preload_libraries = 'pg_stat_statements';

For the changes to take effect, you will have to restart your database. You can do that by typing exit in the terminal to stop your current session.

Here's how you can restart the PostgreSQL server:

sudo -u postgres [comando]


In [38]:
#sudo -u postgres /Library/PostgreSQL/16/bin/pg_ctl restart -D /Library/PostgreSQL/16/data


Terminal CLI

psql --username=postgres --host=localhost 

\connect demo

\dx

|        Name        | Version |   Schema   |                              Description                               |
|--------------------|---------|------------|------------------------------------------------------------------------|
| pg_stat_statements |   1.10  |  bookings  | track planning and execution statistics of all SQL statements executed |
| plpgsql            |   1.0   | pg_catalog | PL/pgSQL procedural language                                            |


Notice how pg_stat_statements has been installed.

You can also check the shared_preload_libraries with:

In [35]:
%sql show shared_preload_libraries;

 * postgresql://postgres:***@localhost:5432/demo
(psycopg2.errors.AdminShutdown) terminating connection due to administrator command
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

[SQL: show shared_preload_libraries;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [36]:
# Especificar los detalles de la conexión
user = 'postgres'
host = 'localhost'
port = '5432'  # Puerto predeterminado de PostgreSQL
dbname = 'demo'  # Nombre de la base de datos

# Crear la URL de conexión
connection_string = f'postgresql://{user}:{password}@{host}:{port}/{dbname}'

# Conectar a la base de datos
%sql $connection_string

In [37]:
%sql show shared_preload_libraries;

 * postgresql://postgres:***@localhost:5432/demo
1 rows affected.


[('pg_stat_statements',)]

pg_stat_statements is also shown under shared_preload_libraries.

Since the results returned by pg_stat_statements can be quite long, let’s turn on expanded table formatting with the following command:

In [39]:
# \x

This will display the output tables in an expanded table format.

demo-# \x

Expanded display is on.

You can turn it off by repeating the \x command.

From the pg_stat_statements documentation, you’ll see the various columns available to be retrieved.

Let’s say you wanted to retrieve the database ID, the query, and total time that it took to execute the statement (in milliseconds).

In [40]:
%sql SELECT dbid, query, total_exec_time FROM pg_stat_statements;

 * postgresql://postgres:***@localhost:5432/demo
10 rows affected.


[(25020, 'select pg_catalog.version()', 0.109166),
 (25020, 'show transaction isolation level', 0.001417),
 (25020, 'show standard_conforming_strings', 0.000625),
 (25020, 'BEGIN', 0.009459),
 (5, 'show shared_preload_libraries', 0.020959),
 (25020, 'show shared_preload_libraries', 0.00675),
 (25020, 'SELECT t.oid, typarray\nFROM pg_type t JOIN pg_namespace ns\n    ON typnamespace = ns.oid\nWHERE typname = $1', 0.184084),
 (25020, 'select current_schema()', 0.004001),
 (25020, 'COMMIT', 0.001),
 (25020, 'ROLLBACK', 0.0006670000000000001)]

Unlike pg_stat_activity, which showed the latest query that was run, pg_stat_statements shows an aggregated view of the queries that were run since the extension was installed.


What if you wanted to check which datbase name matches the database ID?

In [41]:
%sql SELECT oid, datname FROM pg_database;

 * postgresql://postgres:***@localhost:5432/demo
6 rows affected.


[(5, 'postgres'),
 (16398, 'coffee'),
 (1, 'template1'),
 (4, 'template0'),
 (16824, 'restored_demo'),
 (25020, 'demo')]

Based on this, you can now see that database ID 16384 is the demo database. This makes sense because you performed the query show shared_preload libraries on the demo database, which appeared in pg_stat_statements.

It’s important to note that adding these extensions can increase your server load, which may affect performance. If you need to drop the extension, you can achieve that with the following command:

In [42]:
# DROP EXTENSION pg_stat_statements;

If you check the current extensions with \dx, you’ll also see that pg_stat_statements no longer appears.

|    Name   | Version |   Schema   |                              Description                               |
|-----------|---------|------------|------------------------------------------------------------------------|
|  plpgsql  |   1.0   | pg_catalog |                      PL/pgSQL procedural language                      |


If you check the current extensions with \dx, you’ll also see that pg_stat_statements no longer appears.

You should also reset the shared_preload_libraries in the configuration file:

In [43]:
# ALTER SYSTEM RESET shared_preload_libraries;

After this, you’ll need to exit the terminal and restart the PostgreSQL CLI to see the changes reflected in show shared_preload_libraries;.

In [45]:
# sudo -u postgres /Library/PostgreSQL/16/bin/pg_ctl restart -D /Library/PostgreSQL/16/data

In [46]:
# Especificar los detalles de la conexión
user = 'postgres'
host = 'localhost'
port = '5432'  # Puerto predeterminado de PostgreSQL
dbname = 'demo'  # Nombre de la base de datos

# Crear la URL de conexión
connection_string = f'postgresql://{user}:{password}@{host}:{port}/{dbname}'

# Conectar a la base de datos
%sql $connection_string

In [47]:
%sql show shared_preload_libraries;

 * postgresql://postgres:***@localhost:5432/demo
1 rows affected.


[('',)]

### Task C: Monitor with pgAdmin

Another method of monitoring your database comes in the form of pgAdmin. In order to use this tool, you’ll need to first launch it.

On the home page, under Dashboard, you will have access to server or database statistics, depending on which you are looking at.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/2c-server_statistics.png" alt="Server Statistics" width="800" height="400">


On the home page, under Dashboard, you will have access to server or database statistics, depending on which you are looking at.

| Chart                     | Description                                                                                                                                                                        |
|---------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Server/Database sessions  | Displays the total sessions that are running. For servers, this is similar to the pg_stat_activity, and for databases, this is similar to the pg_stat_database.                |
| Transactions per second   | Displays the commits, rollbacks, and transactions taking place.                                                                                                                    |
| Tuples in                 | Displays the number of tuples (rows) that have been inserted, updated, and deleted, similar to the tup_inserted, tup_updated, and tup_deleted columns from pg_stat_database. |
| Tuples out                | Displays the number of tuples (rows) that have been fetched (returned as output) or returned (read or scanned). This is similar to tup_fetched and tup_returned from pg_stat_database. |
| Server activity           | Displays the sessions, locks, prepared transactions, and configuration for the server. In the Sessions tab, it offers a look at the breakdown of the sessions that are currently active on the server, similar to the view provided by pg_stat_activity. To check for any new processes, you can select the refresh button at the top-right corner. |


You can test these charts out by starting another session.

Return to the CLI environment.  


Once you have started that instance, switch back to the tab with pgAdmin.

What do you notice?

Consider this: Which chart(s) monitors active sessions? Remember that one of the charts may need to be refreshed before updates are shown.

You may have noticed that the Server sessions saw an increase of sessions. It increased from 7 to 8 sessions. This makes sense since you started a new session with PostgreSQL CLI.

To see that change reflected in Server Activity, you’ll have to click the refresh button to see that an additional postgres database session appeared.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/2c-view_dashboard_changes.png" alt="Dashboard Changes" width="800" height="400">


To see the dashboard for the demos database, navigate to the left panel and select the Databases dropdown and then select the demo database to connect to it.

As you can see, similar statistics are displayed for the database.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/2c-database_statistics.png" alt="Database Statistics" width="800" height="400">


Let’s run a query on the database! To do that, navigate to the menu bar and select Tools > Query Tool.

You can run any query. To keep things simple, let’s run the following to select all the data from the bookings table:

SELECT * FROM bookings;

Select the run button. You will see that the query has successfully loaded.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/2c-select_all_bookings.png" alt="Select All Bookings" width="800" height="400">


In pgAdmin, switch back to the database’s Dashboard tab. You can refresh the Server activity and check to see if any of the charts have shown a spike since the data was retrieved.

Recall what you queried. Which chart would reflect those changes?

You may have noticed that the number of tuples (rows) returned (read/scanned) was greater than 250,000.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/2c-database_dashboard_query.png" alt="Database Dashboard Query" width="800" height="400">


You can check the number of rows scanned with **EXPLAIN**:

In [2]:
# EXPLAIN SELECT * FROM bookings;

If you can’t see the full text, simply drag the QUERY PLAN column out.

This statement reveals that 262,788 rows were scanned, which is similar to the amount that was read/scanned based on the spike in the Tuples out chart.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/2c-explain_select_query.png" alt="Explain Select Query" width="800" height="400">


While you can monitor your database through the command line alone, tools like pgAdmin can be helpful in providing a visual representation of how your server and its databases are performing.

PostgreSQL also offers logging capabilities to monitor and troubleshoot your lab, which will be further discussed in the Troubleshooting lab.

### Exercise 3: Optimize Your Database

Data optimization is the maximization of the speed and efficiency of retrieving data from your database. Optimizing your database will improve its performance, whether that’s inserting or retrieving data from your database. Doing this will improve the experience of anyone interacting with the database.

Similar to MySQL, there are optimal data types and maintenance (otherwise known as “vacuuming”) that can be applied to optimize databases.



### Task A: Optimize Data Types
When it comes to optimizing data types, understanding the data values will help in selecting the proper data type for the column.

Let’s take a look at an example in the demo database.

Return to the CLI session that you opened previously (or open a new session if it has been closed).

If you’re no longer conected to the demo database, you can reconnect to it!

\connect demo

\dt

|  Schema  |      Name       | Type  |  Owner   |
|----------|-----------------|-------|----------|
| bookings | aircrafts_data  | table | postgres |
| bookings | airports_data   | table | postgres |
| bookings | boarding_passes | table | postgres |
| bookings | bookings        | table | postgres |
| bookings | flights         | table | postgres |
| bookings | seats           | table | postgres |
| bookings | ticket_flights  | table | postgres |
| bookings | tickets         | table | postgres |


Now that you know which tables are in the database, select the first one, aircrafts_data and see what data you can pull from it. How can you select all of its data?

Which statement can you use to select all the data in this table?

You can use the following query to select all the data from aircrafts_data:

SELECT * FROM aircrafts_data;

| aircraft_code |        model        | range |
|---------------|---------------------|-------|
|     773       | Boeing 777-300      | 11100 |
|     763       | Boeing 767-300      |  7900 |
|     SU9       | Sukhoi Superjet-100 |  3000 |
|     320       | Airbus A320-200     |  5700 |
|     321       | Airbus A321-200     |  5600 |
|     319       | Airbus A319-100     |  6700 |
|     733       | Boeing 737-300      |  4200 |
|     CN1       | Cessna 208 Caravan  |  1200 |
|     CR2       | Bombardier CRJ-200  |  2700 |



You can see that there are 9 entries in total with three columns: aircraft_code, model, and range.



For the purposes of this lab, we’ll create a hypothetical situation that will potentially require changing the data types of columns to optimize them.

Let’s say that aircraft_code is always set to three characters, model will always be in a JSON format and range has a maximum value of 12,000 and minimum value of 1,000.

In this case, what would be the best data types for each column?

Take a look at the Data Types Documentation by PostgreSQL to see which data types would fit the columns!

Based on the documentation, the following data types would be suitable for the following columns:

* **aircraft_code:** char(3), since you know that the aircraft code will always be fixed to three characters.
* **model:** json, which is a special data type that PostgreSQL supports.
* **range:** smallint, since the range of its numbers falls between -32,768 to 32,767.

You can check the current data types (and additional details such as the indexes and constraints) of the aircrafts_data table with the following:

\d aircrafts_data

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/3a-detailed_output_aircraft_data.png" alt="Detailed Output Aircraft Data" width="800" height="150">


Notice that most of the columns in this table have been optimized for our sample scenario, except for the range. This may be because the range was unknown in the original database.

For this lab, let’s take the opportunity to optimize that column for your hypothetical situation. You can do this by changing the data type of the column.

Please note that in this lab you’ll first need to drop a view, which is another way our data can be presented, in order to change the column’s data type. Otherwise, you will encounter an error. This is a special case for this database because you loaded a SQL file that included commands to create views. In your own database, you may not need to drop a view.

To drop the aircrafts view, use the following command:

In [3]:
#DROP VIEW aircrafts;

To change the column’s data type, you’ll use the following command:

In [4]:
#ALTER TABLE aircrafts_data ALTER COLUMN range TYPE smallint;

aircrafts_data is the table you want to change and range is the column you want to change to data type smallint.

Now, let’s check the table’s columns and data types again!

In [5]:
# \d aircrafts_data

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/3a-modified_datatype_aircraftdata.png" alt="Modified Datatype Aircraft Data" width="800" height="150">


You can see that the data type has successfully been changed, optimizing your table in this hypothetical situation.

### Task B: Vacuum Your Databases

In your day-to-day life, you can vacuum our rooms to keep them neat and tidy. You can do the same with databases by maintaining and optimizing them with some vacuuming.

In PostgreSQL, vacuuming means to clean out your databases by reclaiming any storage from “dead tuples”, otherwise known as rows that have been deleted but have not been cleaned out.

Generally, the autovacuum feature is automatically enabled, meaning that PostgreSQL will automate the vacuum maintenance process for you.

You can check if this is enabled with the following command:

In [6]:
# show autovacuum;

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/3b-enabled_autovacuum.png" alt="Enabled Autovacuum" width="800" height="100">


As you can see, autovacuum is enabled.

Since autovacuum is enabled, let’s check to see when your database was last vacuumed.

To do that, you can use the pg_stat_user_tables, which displays statistics about each table that is a user table (instead of a system table) in the database. The columns that are returned are the same ones listed in pg_stat_all_tables documentation.

What if you wanted to check the table (by name), the estimated number of dead rows that it has, the last time it was autovacuumed, and how many times it has been autovacuumed?



Recall how you can select specific columns from statistics.

To select the table name, number of dead rows, the last time it was autovacuumed, and the number of times this table has been autovacuumed, you can use the following query:

In [7]:
# SELECT relname, n_dead_tup, last_autoanalyze, autovacuum_count FROM pg_stat_user_tables;

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DB0231EN-SkillsNetwork/labs/PostgreSQL/Lab%20-%20Monitoring%20and%20Optimizing%20your%20Databases%20in%20PostgreSQL/images/3b-check_pgstatusertables_statistics.png" alt="Check pgstattables Statistics" width="800" height="150">


Notice that you currently don’t have any “dead tuples” (deleted rows that haven’t yet been cleaned out) and so far, these tables have been autovacuumed once. This makes sense given that the database was just created and based on the logs, autovacuumed then.

# Conclusion
Congratulations! Now, not only do you know how to monitor and optimize your database with the CLI, but you can also do so with pgAdmin. You will now be able to apply this knowledge to any PostgreSQL databases you create and modify in the future.